src/hooks/useWalletsHiddenBalances.ts

```diff
@@ -51,57 +54,65 @@ const useWalletsHiddenBalances = (wallets: AllRainbowWallets): WalletBalanceResu
   );
 
   const calculateHiddenBalanceForAddress = useCallback(
-    (address: Address) => {
+    (address: Address, data?: UserAssetsResult) => {
       const lowerCaseAddress = address.toLowerCase() as Address;
-      const hiddenAssetBalance = getHiddenAssetBalance({ address, nativeCurrency, connectedToHardhat });
+      const hiddenAssetBalance = getHiddenAssetBalance({ address, nativeCurrency, connectedToHardhat, data });
 
-      console.log('calculateHiddenBalanceForAddress', address, hiddenAssetBalance);
-      setHiddenBalances(prev => ({
-        ...prev,
-        [lowerCaseAddress]: hiddenAssetBalance,
-      }));
+      setHiddenBalances(prev => {
+        const newBalance = hiddenAssetBalance;
+        if (!prev[lowerCaseAddress] || !isEqual(prev[lowerCaseAddress], newBalance)) {
+          return {
+            ...prev,
+            [lowerCaseAddress]: newBalance,
+          };
+        }
+        return prev;
+      });
     },
     [nativeCurrency, connectedToHardhat]
   );
 
   useEffect(() => {
-    allAddresses.forEach(calculateHiddenBalanceForAddress);
+    allAddresses.forEach(address => {
+      calculateHiddenBalanceForAddress(address);
+    });
   }, [allAddresses, calculateHiddenBalanceForAddress]);
 
-  // TODO: This is not working as intended. .subscribe() does not trigger when the value changes.
   useEffect(() => {
-    console.log('Setting up subscriptions for addresses:', allAddresses);
+    const assetSubscriptions = allAddresses.map(address => {
+      return queryClient.getQueryCache().subscribe(event => {
+        if (
+          _isEqual(event.query.queryKey, userAssetsQueryKey({ address, currency: nativeCurrency, connectedToHardhat })) &&
+          event.query.isStale()
```

Comment:
why does the query need to be stale? i think if the cache is updated at all you would want to reflect those changes here